ta/pkcs11/src/persistent_token.c

```diff
@@ -45,82 +45,82 @@ static TEE_Result open_db_file(struct ck_token *token,
 					out_hdl);
 }
 
-static TEE_Result get_pin_file_name(struct ck_token *token,
-				    enum pkcs11_user_type user,
-				    char *name, size_t size)
+void update_persistent_db(struct ck_token *token)
 {
-	int n = snprintf(name, size,
-			 "token.db.%u-pin%d", get_token_id(token), user);
-
-	if (n < 0 || (size_t)n >= size)
-		return TEE_ERROR_SECURITY;
-	else
-		return TEE_SUCCESS;
-}
-
-static TEE_Result open_pin_file(struct ck_token *token,
-				enum pkcs11_user_type user,
-				TEE_ObjectHandle *out_hdl)
-{
-	char file[PERSISTENT_OBJECT_ID_LEN] = { };
 	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_ObjectHandle db_hdl = TEE_HANDLE_NULL;
 
-	res = get_pin_file_name(token, user, file, sizeof(file));
-	if (res)
-		return res;
+	res = open_db_file(token, &db_hdl);
+	if (res) {
+		EMSG("Failed to open token persistent db: %#"PRIx32, res);
+		TEE_Panic(0);
+	}
+	res = TEE_WriteObjectData(db_hdl, token->db_main,
+				  sizeof(*token->db_main));
+	if (res) {
+		EMSG("Failed to write to token persistent db: %#"PRIx32, res);
+		TEE_Panic(0);
+	}
 
-	return TEE_OpenPersistentObject(TEE_STORAGE_PRIVATE, file, sizeof(file),
-					0, out_hdl);
+	TEE_CloseObject(db_hdl);
 }
 
-static void init_pin_keys(struct ck_token *token, enum pkcs11_user_type user)
+static enum pkcs11_rc do_hash(uint32_t user, const uint8_t *pin,
+			      size_t pin_size, uint32_t salt,
+			      uint8_t hash[TEE_MAX_HASH_SIZE])
 {
-	TEE_Result res = TEE_ERROR_GENERIC;
-	TEE_ObjectHandle key_hdl = TEE_HANDLE_NULL;
+	TEE_Result res = TEE_SUCCESS;
+	TEE_OperationHandle oh = TEE_HANDLE_NULL;
+	uint32_t sz = TEE_MAX_HASH_SIZE;
 
-	res = open_pin_file(token, user, &key_hdl);
-
-	if (res == TEE_SUCCESS)
-		DMSG("PIN key found");
+	res = TEE_AllocateOperation(&oh, TEE_ALG_SHA256, TEE_MODE_DIGEST, 0);
+	if (res)
+		return tee2pkcs_error(res);
 
-	if (res == TEE_ERROR_ITEM_NOT_FOUND) {
-		TEE_Attribute attr = { };
-		TEE_ObjectHandle hdl = TEE_HANDLE_NULL;
-		uint8_t pin_key[16] = { };
-		char file[PERSISTENT_OBJECT_ID_LEN] = { };
+	TEE_DigestUpdate(oh, &user, sizeof(user));
+	TEE_DigestUpdate(oh, &salt, sizeof(salt));
+	res = TEE_DigestDoFinal(oh, pin, pin_size, hash, &sz);
+	TEE_FreeOperation(oh);
 
-		TEE_MemFill(&attr, 0, sizeof(attr));
+	if (res)
+		return PKCS11_CKR_GENERAL_ERROR;
 
-		TEE_GenerateRandom(pin_key, sizeof(pin_key));
-		TEE_InitRefAttribute(&attr, TEE_ATTR_SECRET_VALUE,
-				     pin_key, sizeof(pin_key));
+	memset(hash + sz, 0, TEE_MAX_HASH_SIZE - sz);
+	return PKCS11_CKR_OK;
+}
 
-		res = TEE_AllocateTransientObject(TEE_TYPE_AES, 128, &hdl);
-		if (res)
-			TEE_Panic(0);
+enum pkcs11_rc hash_pin(enum pkcs11_user_type user, const uint8_t *pin,
+			size_t pin_size, uint32_t *salt,
+			uint8_t hash[TEE_MAX_HASH_SIZE])
+{
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	uint32_t s = 0;
 
-		res = TEE_PopulateTransientObject(hdl, &attr, 1);
-		if (res)
-			TEE_Panic(0);
+	TEE_GenerateRandom(&s, sizeof(s));
+	if (!s)
+		s++;
```

Comment:
A salt of `0` indicates that there is no salt set AFAIR from the rest of the code. I think the bias would be small, but agree that your implementation for this would be cleaner. Do you want to create a PR against current master with your fixes?

Reply:
From reading the rest of the code a salt of 0 indicates the user PIN has not yet been initialised.